File: neetcode150/06_linked_list/0019_remove_nth_node_from_end_of_list.cpp
/*
 * =====================================================================
 * LeetCode 19 : Remove Nth Node From End of List               (Medium)
 * https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 * Category   : Linked List
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given the head of a list and n, remove the n-th node COUNTING
 *   FROM THE END (1-indexed) and return the head. The list is singly
 *   linked — there are no back pointers to locate the tail cheaply.
 *       1→2→3→4→5, n=2  →  1→2→3→5
 * ---------------------------------------------------------------------
 * INTUITION
 *   "n-th from the end" = "n steps BEFORE the null at the end".
 *   Two pointers hold a measuring stick of length n: if right is
 *   already n steps ahead of left, then when right reaches null,
 *   left is EXACTLY at the node before the victim. Deleting = skip.
 * ---------------------------------------------------------------------
 * APPROACH 1 — One pass: dummy + fixed-gap two pointers (BEST)
 *   dummy → head. right walks n steps ahead of left (they share the
 *   gap). Walk both until right hits null → left->next is the victim;
 *   left->next = left->next->next deletes it. The dummy guarantees
 *   the victim may be the original head without an edge case.
 *   Time  Complexity : O(n)  — one pass.
 *   Space Complexity : O(1)  — two pointers.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Two passes: count, then delete
 *   Pass 1 counts L nodes (O(L)). The victim is the (L − n + 1)-th
 *   node from the start; walk to its predecessor (L − n steps from
 *   head) and skip it. Needs the dummy for the n == L case too.
 *   Time  Complexity : O(n)  — but two passes.
 *   Space Complexity : O(1).
 * ---------------------------------------------------------------------
 * DEEP DIVE — deleting a node = pointer surgery
 *   A singly linked list can only delete a node if you can SEE its
 *   predecessor — that is why we stop at left->next, never at the
 *   victim itself. The one tricky case is deleting the head (no
 *   predecessor exists): the dummy node makes the head a normal
 *   node with a normal predecessor. This "dummy absorbs the edge
 *   case" idea is the same trick that powers 0021 (merge) and
 *   0025 (reverse k-group). Memory hygiene note: in a real runtime
 *   you also free the skipped node (we do, then rebuild from test
 *   data). Two-pointer gap windows appear in real systems: network
 *   ring buffers keep a read pointer exactly K bytes behind a write
 *   pointer, i.e. "the n-from-the-end element".
 * =====================================================================
 */

#include <vector>          // std::vector — test data
#include <chrono>          // std::chrono — timing
#include <iostream>        // std::cout — output
#include <sys/resource.h>  // getrusage — peak RAM
#include "list_node.h"     // our header: ListNode, buildList

/* =====================================================================
 * APPROACH 1 : One pass with a dummy node + fixed-gap pointers
 * ===================================================================== */
class Solution_OnePass {
public:
    /*
     * Purpose : Delete the n-th node from the end in a single walk.
     * Inputs  : head — first node of the list; n — 1-indexed distance
     *                  from the end (guaranteed ≤ list length).
     * Output  : Head of the list with the target node removed.
     */
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Dummy so the original head can be deleted like any node.
        ListNode* dummy = new ListNode(0, head);
        ListNode* left = dummy;
        ListNode* right = dummy;

        // Stretch the measuring stick: right jumps n steps first.
        for (int i = 0; i < n; ++i) {
            right = right->next;
        }

        // Slide the stick: gap stays n; when right is at null,
        // left sits exactly at the node BEFORE the victim.
        while (right->next != nullptr) {
            left = left->next;
            right = right->next;
        }

        // Pointer surgery: skip the victim (and free it).
        ListNode* victim = left->next;
        left->next = victim->next;      // bridge over the deleted node
        delete victim;                  // return its memory

        ListNode* newHead = dummy->next;
        delete dummy;                   // scaffold goes away
        return newHead;
    }
};

/* =====================================================================
 * APPROACH 2 : Two passes — count the length, then delete
 * ===================================================================== */
class Solution_TwoPass {
public:
    /*
     * Purpose : Delete the n-th node from the end using two walks.
     * Inputs  : head — first node of the list; n — 1-indexed distance
     *                  from the end.
     * Output  : Head of the list with the target node removed.
     */
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Pass 1: measure the total length L.
        int length = 0;
        for (ListNode* cur = head; cur != nullptr; cur = cur->next) {
            ++length;
        }

        // Victim from the start is (L − n + 1); we need its
        // predecessor, i.e. the (L − n)-th node. Dummy handles n == L.
        ListNode* dummy = new ListNode(0, head);
        ListNode* cur = dummy;
        for (int i = 0; i < length - n; ++i) {
            cur = cur->next;            // walk to the predecessor
        }

        ListNode* victim = cur->next;
        cur->next = victim->next;       // bridge over it
        delete victim;

        ListNode* newHead = dummy->next;
        delete dummy;
        return newHead;
    }
};

/* =====================================================================
 * BENCHMARK — 3 tests, both approaches, time + peak RAM
 * ===================================================================== */
int main() {
    std::vector<std::vector<int>> vals = {{1, 2, 3, 4, 5}, {1}, {1, 2}};
    std::vector<int> n = {2, 1, 2};
    std::vector<std::vector<int>> exp = {{1, 2, 3, 5}, {}, {2}};

    auto verify = [&](ListNode* h, std::size_t t) {
        for (int want : exp[t]) {
            if (h == nullptr || h->val != want) return false;
            h = h->next;
        }
        return h == nullptr;
    };

    bool ok1 = true, ok2 = true;

    // ---- Approach 1 ----
    struct rusage r1a, r1b;
    getrusage(RUSAGE_SELF, &r1a);
    auto s1 = std::chrono::high_resolution_clock::now();
    for (std::size_t t = 0; t < vals.size(); ++t) {
        ListNode* head = buildList(vals[t]);
        ListNode* out = Solution_OnePass().removeNthFromEnd(head, n[t]);
        ok1 = verify(out, t) && ok1;
        deleteList(out);
    }
    auto e1 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r1b);

    // ---- Approach 2 ----
    struct rusage r2a, r2b;
    getrusage(RUSAGE_SELF, &r2a);
    auto s2 = std::chrono::high_resolution_clock::now();
    for (std::size_t t = 0; t < vals.size(); ++t) {
        ListNode* head = buildList(vals[t]);
        ListNode* out = Solution_TwoPass().removeNthFromEnd(head, n[t]);
        ok2 = verify(out, t) && ok2;
        deleteList(out);
    }
    auto e2 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r2b);

    double us1 = std::chrono::duration_cast<std::chrono::nanoseconds>(e1 - s1).count() / 1000.0;
    double us2 = std::chrono::duration_cast<std::chrono::nanoseconds>(e2 - s2).count() / 1000.0;
    long kb1 = (r1b.ru_maxrss - r1a.ru_maxrss) / 1024;
    long kb2 = (r2b.ru_maxrss - r2a.ru_maxrss) / 1024;

    std::cout << "0019 Remove Nth Node From End of List\n";
    std::cout << "Approach 1 (one pass) : time = " << us1 << " µs  peak-ram-delta = " << kb1 << " KB\n";
    std::cout << "Approach 2 (two pass) : time = " << us2 << " µs  peak-ram-delta = " << kb2 << " KB\n";
    std::cout << (ok1 && ok2 ? "PASS : both approaches removed the right node on all 3 tests."
                             : "FAIL : at least one test produced a wrong list.")
              << "\n";
    return 0;
}

File: neetcode150/06_linked_list/0019_remove_nth_node_from_end_of_list.py
"""
=====================================================================
LeetCode 19 : Remove Nth Node From End of List               (Medium)
https://leetcode.com/problems/remove-nth-node-from-end-of-list/
Category   : Linked List
---------------------------------------------------------------------
PROBLEM
    Remove the n-th node counting from the END (1-indexed) of a
    singly linked list; return the head.
        1→2→3→4→5, n=2  →  1→2→3→5
---------------------------------------------------------------------
INTUITION
    "n-th from the end" = "n steps before the null at the end".
    Two pointers holding a measuring stick of length n: when the
    right one reaches None, the left one is at the node BEFORE
    the victim. Deleting = skipping (pointer surgery).
---------------------------------------------------------------------
APPROACH 1 — One pass: dummy + fixed-gap two pointers (BEST)
    right walks n steps ahead; both slide until right hits None;
    left.next is the victim; left.next = left.next.next deletes.
    Time  Complexity : O(n)   Space Complexity : O(1)
---------------------------------------------------------------------
APPROACH 2 — Two passes: count, then delete
    Pass 1 counts L nodes; victim is node (L − n + 1) from start;
    walk to its predecessor and skip it.
    Time  Complexity : O(n)   Space Complexity : O(1)
---------------------------------------------------------------------
REAL LIFE
    • Ring buffers keep a read pointer K bytes behind a write
      pointer — the same fixed-gap window.
    • "Undo N steps" trackers and sliding-window log tails.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import Optional


class ListNode:
    """Minimal singly linked list node used by every test below."""

    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


def build_list(values: list[int]) -> Optional[ListNode]:
    """Turn a Python list into a linked list; returns the head."""
    dummy = ListNode()
    tail = dummy
    for v in values:
        tail.next = ListNode(v)
        tail = tail.next
    return dummy.next


def to_list(head: Optional[ListNode]) -> list[int]:
    """Walk a linked list back into a Python list (for verification)."""
    out: list[int] = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


# =====================================================================
# APPROACH 1 : One pass — dummy + fixed-gap two pointers (best)
# =====================================================================
class Solution_OnePass:
    """
    Purpose : Delete the n-th node from the end in a single walk.
    Inputs  : head — first node of the list; n — 1-indexed distance
                    from the end (guaranteed ≤ list length).
    Output  : Head of the list with the target node removed.
    """

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)         # absorbs the delete-head case
        left: Optional[ListNode] = dummy
        right: Optional[ListNode] = dummy

        # Stretch the measuring stick: right jumps n steps first.
        for _ in range(n):
            right = right.next if right else None

        # Slide: gap stays n; right at None → left before the victim.
        while right is not None and right.next is not None:
            left = left.next if left else None
            right = right.next

        # Pointer surgery: bridge over the victim.
        if left is not None and left.next is not None:
            left.next = left.next.next

        return dummy.next


# =====================================================================
# APPROACH 2 : Two passes — count the length, then delete
# =====================================================================
class Solution_TwoPass:
    """
    Purpose : Delete the n-th node from the end using two walks.
    Inputs  : head — first node of the list; n — 1-indexed distance
                    from the end.
    Output  : Head of the list with the target node removed.
    """

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Pass 1: measure the total length L.
        length = 0
        cur = head
        while cur is not None:
            length += 1
            cur = cur.next

        # Victim is (L − n + 1)-th from start; walk to its
        # predecessor: the (L − n)-th node. Dummy handles n == L.
        dummy = ListNode(0, head)
        cur = dummy
        for _ in range(length - n):
            cur = cur.next if cur else None

        if cur is not None and cur.next is not None:
            cur.next = cur.next.next       # skip the victim

        return dummy.next


# =====================================================================
# BENCHMARK — 3 tests, both approaches, time + peak memory
# =====================================================================
if __name__ == "__main__":
    tests: list[tuple[list[int], int]] = [
        ([1, 2, 3, 4, 5], 2),
        ([1], 1),
        ([1, 2], 2),
    ]
    expected: list[list[int]] = [[1, 2, 3, 5], [], [2]]

    # --- Approach 1 ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok1 = all(
        to_list(Solution_OnePass().removeNthFromEnd(build_list(v), n)) == e
        for (v, n), e in zip(tests, expected)
    )
    t1 = time.perf_counter()
    _, peak1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2 ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok2 = all(
        to_list(Solution_TwoPass().removeNthFromEnd(build_list(v), n)) == e
        for (v, n), e in zip(tests, expected)
    )
    t1b = time.perf_counter()
    _, peak2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print("0019 Remove Nth Node From End of List")
    print(f"Approach 1 (one pass) : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (two pass) : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both approaches removed the right node on all 3 tests."
          if ok1 and ok2 else "FAIL : at least one test produced a wrong list.")
